File: javascript/src/0028.js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
function strStr(haystack, needle) {
  let n = haystack.length
  let m = needle.length
  if (m === 0) return 0
  let next = new Int32Array(m)
  for (let i = 1, j = 0; i < m; i++) {
    while (j > 0 && needle[i] !== needle[j]) j = needle[j - 1]
    if (needle[i] === needle[j]) j++
    next[i] = j
  }
  for (let i = 0, j = 0; i < n; i++) {
    while (j > 0 && haystack[i] != needle[j]) j = needle[j - 1]
    if (haystack[i] === needle[j]) j++
    if (j === m) return i - m + 1
  }
  return -1
}

{
  const { strictEqual } = require('assert')

  const CASES = [
    ['hello', 'll', 2],
    ['aaaaa', 'bba', -1],
  ]

  for (const [haystack, needle, excepted] of CASES) {
    const actual = strStr(haystack, needle)
    strictEqual(actual, excepted)
  }

  console.time()
  for (const [haystack, needle, _] of CASES) {
    for (let i = 0; i < 100000; i++) {
      strStr(haystack, needle)
    }
  }
  console.timeEnd()
}

File: python/src/0028.py
class Solution:

    def strStr(self, haystack: str, needle: str) -> int:
        n = len(haystack)
        m = len(needle)
        if m == 0:
            return 0
        next = bytearray(m)
        j = 0
        for i in range(1, m):
            while j > 0 and needle[i] != needle[j]:
                j = next[j - 1]
            if needle[i] == needle[j]:
                j += 1
            next[i] = j
        j = 0
        for i in range(n):
            while j > 0 and haystack[i] != needle[j]:
                j = needle[j - 1]
            if haystack[i] == needle[j]:
                j += 1
            if j == m:
                return i - m + 1
        return -1


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        ('hello', 'll', 2),
        ('aaaaa', 'bba', -1),
    )

    class Test(unittest.TestCase):

        def test(self):
            strStr = Solution().strStr
            for haystack, needle, excepted in CASES:
                actual = strStr(haystack, needle)
                self.assertEqual(actual, excepted)

    unittest.main()

    strStr = Solution().strStr
    for haystack, needle, excepted in CASES:
        print(timeit(lambda: strStr(haystack, needle), number=100000))
